plugins/modules/packaging/os/pacman.py

```diff
@@ -180,357 +182,465 @@
-
-    module.exit_json(changed=False, msg="package(s) already absent")
-
-
-def install_packages(module, pacman_path, state, packages, package_files):
-    install_c = 0
-    package_err = []
-    message = ""
-    data = []
-    diff = {
-        'before': '',
-        'after': '',
-    }
-
-    to_install_repos = []
-    to_install_files = []
-    for i, package in enumerate(packages):
-        # if the package is installed and state == present or state == latest and is up-to-date then skip
-        installed, updated, latestError = query_package(module, pacman_path, package, state)
-        if latestError and state == 'latest':
-            package_err.append(package)
-
-        if installed and (state == 'present' or (state == 'latest' and updated)):
-            continue
-
-        if package_files[i]:
-            to_install_files.append(package_files[i])
+
+class Pacman(object):
+    def __init__(self, module):
+        self.m = module
+
+        self.m.run_command_environ_update = dict(LC_ALL="C")
+        p = self.m.params
+
+        self._msgs = []
+        self._stdouts = []
+        self._stderrs = []
+        self.changed = False
+        self.exit_params = {}
+
+        self.pacman_path = self.m.get_bin_path(p["executable"], True)
+        self.inventory = self._build_inventory()
+
+        # Normalize for old configs
+        if p["state"] == "installed":
+            self.target_state = "present"
+        elif p["state"] == "removed":
+            self.target_state = "absent"
         else:
-            to_install_repos.append(package)
+            self.target_state = p["state"]
+
+    def add_exit_infos(self, msg=None, stdout=None, stderr=None):
+        if msg:
+            self._msgs.append(msg)
+        if stdout:
+            self._stdouts.append(stdout)
+        if stderr:
+            self._stderrs.append(stderr)
+
+    def _set_mandatory_exit_params(self):
+        msg = "\n".join(self._msgs)
+        stdouts = "\n".join(self._stdouts)
+        stderrs = "\n".join(self._stderrs)
+        if stdouts:
+            self.exit_params["stdout"] = stdouts
+        if stderrs:
+            self.exit_params["stderr"] = stderrs
+        self.exit_params["msg"] = msg  # mandatory, but might be empty
+
+    def fail(self, msg=None, stdout=None, stderr=None, **kwargs):
+        self.add_exit_infos(msg, stdout, stderr)
+        self._set_mandatory_exit_params()
+        if kwargs:
+            self.exit_params.update(**kwargs)
+        self.m.fail_json(**self.exit_params)
+
+    def success(self):
+        self._set_mandatory_exit_params()
+        self.m.exit_json(changed=self.changed, **self.exit_params)
+
+    def run(self):
+        if self.m.params["update_cache"]:
+            self.update_package_db()
+
+            if not (self.m.params["name"] or self.m.params["upgrade"]):
+                self.success()
+
+        if self.m.params["upgrade"]:
```

Comment:
Shouldn't this only be called here?
```suggestion
        self.inventory = self._build_inventory()

        if self.m.params["upgrade"]:
```
Otherwise `self.inventory` contains outdated information if `update_cache=true` was specified.

Reply:
Definitely!